File: android/Reguerta/app/src/main/java/com/reguerta/user/domain/products/Product.kt
package com.reguerta.user.domain.products

data class Product(
    val id: String,
    val vendorId: String,
    val companyName: String,
    val name: String,
    val description: String,
    val productImageUrl: String?,
    val price: Double,
    val pricingMode: ProductPricingMode = ProductPricingMode.FIXED,
    val unitName: String,
    val unitAbbreviation: String?,
    val unitPlural: String,
    val unitQty: Double,
    val packContainerName: String?,
    val packContainerAbbreviation: String?,
    val packContainerPlural: String?,
    val packContainerQty: Double?,
    val isAvailable: Boolean,
    val stockMode: ProductStockMode,
    val stockQty: Double?,
    val isEcoBasket: Boolean,
    val isCommonPurchase: Boolean,
    val commonPurchaseType: CommonPurchaseType?,
    val archived: Boolean,
    val createdAtMillis: Long,
    val updatedAtMillis: Long,
    val weightStep: Double? = null,
    val minWeight: Double? = null,
    val maxWeight: Double? = null,
) {
    val isVisibleInOrdering: Boolean
        get() = !archived && isAvailable
}

val Product.effectiveWeightStep: Double
    get() = weightStep?.takeIf { it > 0.0 } ?: unitQty.takeIf { it > 0.0 } ?: 1.0

val Product.minimumSelectionCount: Int
    get() = if (pricingMode == ProductPricingMode.WEIGHT) {
        kotlin.math.ceil((minWeight ?: effectiveWeightStep) / effectiveWeightStep).toInt().coerceAtLeast(1)
    } else {
        1
    }

val Product.maximumSelectionCount: Int?
    get() = if (pricingMode == ProductPricingMode.WEIGHT) {
        maxWeight?.let {
            kotlin.math.floor(it / effectiveWeightStep).toInt().coerceAtLeast(minimumSelectionCount)
        }
    } else {
        null
    }

fun Product.selectedQuantity(selectionCount: Int): Double =
    if (pricingMode == ProductPricingMode.WEIGHT) {
        selectionCount * effectiveWeightStep
    } else {
        selectionCount.toDouble()
    }

enum class ProductPricingMode {
    FIXED,
    WEIGHT,
}

enum class ProductStockMode {
    FINITE,
    INFINITE,
}

enum class CommonPurchaseType {
    SEASONAL,
    SPOT,
}

File: ios/Reguerta/Reguerta/Domain/Products/Product.swift
import Foundation

struct Product: Identifiable, Equatable {
    let id: String
    let vendorId: String
    let companyName: String
    let name: String
    let description: String
    let productImageUrl: String?
    let price: Double
    let pricingMode: ProductPricingMode
    let unitName: String
    let unitAbbreviation: String?
    let unitPlural: String
    let unitQty: Double
    let packContainerName: String?
    let packContainerAbbreviation: String?
    let packContainerPlural: String?
    let packContainerQty: Double?
    let isAvailable: Bool
    let stockMode: ProductStockMode
    let stockQty: Double?
    let isEcoBasket: Bool
    let isCommonPurchase: Bool
    let commonPurchaseType: CommonPurchaseType?
    let archived: Bool
    let createdAtMillis: Int64
    let updatedAtMillis: Int64
    var weightStep: Double?
    var minWeight: Double?
    var maxWeight: Double?

    var isVisibleInOrdering: Bool {
        !archived && isAvailable
    }
}

extension Product {
    var effectiveWeightStep: Double {
        if let weightStep, weightStep.isFinite, weightStep > 0 { return weightStep }
        return unitQty.isFinite && unitQty > 0 ? unitQty : 1
    }

    var minimumSelectionCount: Int {
        guard pricingMode == .weight else { return 1 }
        return max(1, boundedProductUnitCount(
            ceil((minWeight ?? effectiveWeightStep) / effectiveWeightStep)
        ))
    }

    var maximumSelectionCount: Int? {
        guard pricingMode == .weight, let maxWeight else { return nil }
        return max(
            minimumSelectionCount,
            boundedProductUnitCount(floor(maxWeight / effectiveWeightStep))
        )
    }

    func selectedQuantity(selectionCount: Int) -> Double {
        pricingMode == .weight ? Double(selectionCount) * effectiveWeightStep : Double(selectionCount)
    }
}

func boundedProductUnitCount(_ value: Double) -> Int {
    guard value.isFinite, value > 0 else { return 0 }
    guard value < Double(Int32.max) else { return Int(Int32.max) }
    return Int(value)
}

enum ProductPricingMode: String, Equatable, Sendable {
    case fixed
    case weight
}

enum ProductStockMode: String, Equatable, Sendable {
    case finite
    case infinite
}

enum CommonPurchaseType: String, Equatable, Sendable, CaseIterable {
    case seasonal
    case spot
}
